### PR TITLE
pkg/*: use Kubernetes ServiceAccount as service identity

### PR DESCRIPTION
**Pull request by shashankram**
**Description**:
This change does the following:

- Uses kubernetes ServiceAccount as service identity for
  proxy service certificates. The service certificate
  CN/SAN will be of the form `<svc-account>.<namespace>.<domain>`.

- Refactors SDS to make it easier to share code in its internal
  implementation.

- Removes the overloaded usage of the certificate's CN for SNI.
  SNI (Server Name Identifier) is based on the service FQDN,
  while the service certificate is derived from the ServiceIdentity
  being used.

- Removes SAN matching for the downstream/upstream service
  in permissive mode, where there are no SMI TrafficTarget policies.
  In permissive mode, every service within the mesh is allowed to
  reach every other service as long as they use the correct certificate.
 
- Refactors sds/response_test.go to make it possible to use
  table-driven testing to test different unit testing scenarios.

Part of #1965

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [X]
- CLI Tool               [ ]
- Certificate Management [X]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [X]
- Tests                  [X]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
`No`

### CODE REVIEW

**Review by snehachhabria — DISMISSED** [review]

**Review by snehachhabria — APPROVED** [review]

File: go.sum [hunk 1]
```diff
@@ -243,6 +243,7 @@ github.com/envoyproxy/go-control-plane v0.9.0/go.mod h1:YTl/9mNaCwkRvm6d1a2C3ymF
 github.com/envoyproxy/go-control-plane v0.9.1-0.20191026205805-5f8ba28d4473/go.mod h1:YTl/9mNaCwkRvm6d1a2C3ymFceY/DCBVvsKhRF0iEA4=
 github.com/envoyproxy/go-control-plane v0.9.6 h1:GgblEiDzxf5ajlAZY4aC8xp7DwkrGfauFNMGdB2bBv0=
 github.com/envoyproxy/go-control-plane v0.9.6/go.mod h1:GFqM7v0B62MraO4PWRedIbhThr/Rf7ev6aHOOPXeaDA=
+github.com/envoyproxy/go-control-plane v0.9.7 h1:EARl0OvqMoxq/UMgMSCLnXzkaXbxzskluEBlMQCJPms=
```
Comment (draychev) [inline go.sum:246]: 🤔

File: pkg/envoy/sds/response.go [hunk 2]
```diff
@@ -23,59 +24,86 @@ var directionMap = map[envoy.SDSCertType]string{
...
+	svcAccount, err := catalog.GetServiceAccountFromProxyCertificate(proxy.GetCommonName())
 	if err != nil {
-		log.Error().Err(err).Msgf("Error issuing a certificate for proxy service %s", serviceForProxy)
+		log.Error().Err(err).Msgf("Error retrieving SerivceAccount for proxy %s", proxy.GetCommonName())
 		return nil, err
 	}
 
-	// Iterate over the list of tasks and create response structs to be
-	// sent to the proxy that made the discovery request
+	sdsImpl := newSDSImpl(proxy, meshCatalog, certManager, cfg, svcList, svcAccount)
```
Comment (draychev) [inline pkg/envoy/sds/response.go:43]: Not sure what `impl` means here or anywhere below really. I would prefer to use something that reads easily in plain english and helps you understand what's behind this variable.

No need to change this PR - but eventually would be good to add clarity so it is easy to read.

File: pkg/envoy/sds/response.go [hunk 3]
```diff
@@ -85,17 +113,17 @@ func getEnvoySDSSecrets(cert certificate.Certificater, proxy *envoy.Proxy, reque
 			continue
 		}
 
-		if serviceForProxy != sdsCert.MeshService {
-			log.Error().Msgf("Proxy %s (service %s) requested service certificate %s; this is not allowed", proxy.GetCommonName(), serviceForProxy, sdsCert.MeshService)
+		if proxyService != sdsCert.MeshService {
+			log.Debug().Msgf("Proxy %s (service %s) requested service certificate %s; this is not allowed", s.proxy.GetCommonName(), proxyService, sdsCert.MeshService)
```
Comment (draychev) [inline pkg/envoy/sds/response.go:117]: `Debug` makes sense, except I worry that it would hide this message, when I believe it should be emitted with a higher level of urgency. If error seems to high - I'd say warn.

**Review by draychev — APPROVED** [review]
Left a few comments that could be addressed in follow-up PRs - feel free to ignore.
Thanks for working on this!

File: pkg/envoy/sds/response.go (hunk 2 above)
Comment (shashankram) [reply pkg/envoy/sds/response.go:43]: I left a comment on what `sdsImpl` refers to in `sds/types.go`.

File: pkg/envoy/sds/response.go (hunk 3 above)
Comment (shashankram) [reply pkg/envoy/sds/response.go:117]: `Warn` seems appropriate for this use case, will address!

**Comment by shashankram** [comment]
> Left a few comments that could be addressed in follow-up PRs - feel free to ignore.
> Thanks for working on this!

Thanks for the review, will address the nits in a follow-up coming shortly.

File: pkg/envoy/sds/response.go (hunk 3 above)
Comment (shashankram) [reply pkg/envoy/sds/response.go:117]: Addressed in https://github.com/openservicemesh/osm/pull/1995
